stonesoup/updater/particle.py

```diff
@@ -70,7 +70,236 @@ def predict_measurement(self, state_prediction, measurement_model=None,
+
+        new_particles = self.resampler.resample(
+            hypothesis.prediction.particles)
+
+        return ParticleStateUpdate(new_particles,
+                                   hypothesis,
+                                   timestamp=hypothesis.measurement.timestamp)
+
+    @lru_cache()
+    def predict_measurement(self, state_prediction, measurement_model=None,
+                            **kwargs):
+
+        if measurement_model is None:
+            measurement_model = self.measurement_model
+
+        new_particles = []
+        for particle in state_prediction.particles:
+            new_state_vector = measurement_model.function(
+                particle.state_vector, noise=0, **kwargs)
+            new_particles.append(
+                RaoBlackwellisedParticle(new_state_vector,
+                                         weight=particle.weight,
+                                         parent=particle.parent,
+                                         model_probabilities=particle.model_probabilities))
+
+        return ParticleMeasurementPrediction(
+            new_particles, timestamp=state_prediction.timestamp)
+
+    @staticmethod
+    def calculate_model_probabilities(particle, position_mapping,
+                                      transition_matrix, model_list, time_interval):
+
+        """Calculates the new model probabilities based
+            on the ones calculated in the previous time step"""
+
+        previous_probabilities = particle.model_probabilities
+
+        denominator_components = []
+        # Loop over the current models m_k
+        for l in range(len(previous_probabilities)):
+            selected_model_sum = 0
+            # Loop over the previous models m_k-1
+            for i, model in enumerate(model_list):
+                # Looks up p(m_k|m_k-1)
+                # Note: if p(m_k|m_k-1) = 0 then p(m_k|x_1:k) = 0
+                transition_probability = transition_matrix[i][l]
+                if transition_probability == 0:
+                    break
+                # Getting required states to apply the model to that state vector
+                parent_required_state_space = particle.parent.state_vector[
+                    np.array(position_mapping[i])]
+
+                # The noiseless application of m_k onto x_k-1
+                mean = model.function(
+                    parent_required_state_space,
+                    time_interval=time_interval, noise=False)
+
+                # Input the indices that were removed previously
+                for j in range(len(particle.state_vector)):
+                    if j not in position_mapping[i]:
+                        mean = np.insert(mean, j, 0)
+                mean = mean.reshape((1, -1))[0]
+
+                prob_position_given_model_and_old_position = multivariate_normal.pdf(
+                    np.transpose(particle.state_vector),
+                    mean=mean,
+                    cov=np.diag([0.75 for i in range(len(mean))])
```

Comment:
Looking at this now @Isaac-JenkinsRA, and wondering about where this covariance comes from? (Sorry, I'm not familiar with paper on this filter).

Reply:
This is concerning as I too am not sure what this arbitrary value of 0.75 is supposed to be. The covariance here should be the noise applied to a particle once it has been acted on by a transition model. Hopefully this helps.